src/tools/csi-kata-directvolume/pkg/directvolume/nodeserver.go

```diff
@@ -0,0 +1,503 @@
+		if err != nil {
+			return nil, fmt.Errorf("unmount target path: %w", err)
+		}
+	}
+
+	// Delete the mount point.
+	// Does not return error for non-existent path, repeated calls OK for idempotency.
+	if err := os.RemoveAll(targetPath); err != nil {
+		return nil, fmt.Errorf("remove target path: %w", err)
+	}
+
+	if err := utils.RemoveDirectVolume(targetPath); err != nil {
+		klog.V(4).Infof("remove direct volume failed.")
+		utils.RemoveDirectVolume(targetPath)
+	}
+
+	klog.Infof("direct volume %s has been cleaned up.", targetPath)
+
+	vol, err := dv.state.GetVolumeByID(volumeID)
+	if err != nil {
+		klog.Warningf("volume id %s not found in volume list, nothing to do.")
+		return &csi.NodeUnpublishVolumeResponse{}, nil
+	}
+
+	if !vol.Published.Has(targetPath) {
+		klog.V(4).Infof("volume %q is not published at %q, nothing to do.", volumeID, targetPath)
+		return &csi.NodeUnpublishVolumeResponse{}, nil
+	}
+
+	vol.Published.Remove(targetPath)
+	if err := dv.state.UpdateVolume(vol); err != nil {
+		return nil, err
+	}
+	klog.Infof("volume %s has been unpublished.", targetPath)
+
+	return &csi.NodeUnpublishVolumeResponse{}, nil
+}
+
+func isDirectVolume(VolumeCtx map[string]string) bool {
+	if VolumeCtx[utils.IsDirectVolume] == "True" {
+		return true
+	}
+
+	return false
+}
+
+func (dv *directVolume) NodeStageVolume(ctx context.Context, req *csi.NodeStageVolumeRequest) (*csi.NodeStageVolumeResponse, error) {
+	klog.V(4).Infof("NodeStageVolumeRequest with request %v", req)
+
+	volumeID := req.GetVolumeId()
+	// Check arguments
+	if len(volumeID) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Volume ID missing in request")
+	}
+	stagingTargetPath := req.GetStagingTargetPath()
+	if stagingTargetPath == "" {
+		return nil, status.Error(codes.InvalidArgument, "Target path missing in request")
+	}
+	if req.GetVolumeCapability() == nil {
+		return nil, status.Error(codes.InvalidArgument, "Volume Capability missing in request")
+	}
+
+	if !isDirectVolume(req.VolumeContext) {
+		return nil, status.Errorf(codes.FailedPrecondition, "volume %q is not direct-volume.", req.VolumeId)
+	}
+
+	// Lock before acting on global state. A production-quality
+	// driver might use more fine-grained locking.
+	dv.mutex.Lock()
+	defer dv.mutex.Unlock()
+	// Create direct block device and mkfs with fsType, default fsType is "ext4".
+
+	capacityInBytes := req.VolumeContext[utils.CapabilityInBytes]
+	devicePath, err := utils.CreateDirectBlockDevice(volumeID, capacityInBytes, dv.config.StoragePath)
+	if err != nil {
+		errMsg := status.Errorf(codes.Internal, "setup storage for volume '%s' failed", volumeID)
+		return &csi.NodeStageVolumeResponse{}, errMsg
+	}
+
+	// /full_path_on_host/VolumeId/
+	deviceUpperPath := filepath.Dir(*devicePath)
+	if canMnt, err := utils.CanDoBindMount(dv.config.safeMounter, stagingTargetPath); err != nil {
+		return nil, err
+	} else if !canMnt {
+		klog.Infof("Cannot mount to target path: %s", stagingTargetPath)
```

Comment:
Is stagingTargetPath already mounted in this case ? The log showing "already mounted" looks better

Reply:
fixed